nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -14,84 +14,158 @@
+    std::bind(&CostmapSubscriber::costmapUpdateCallback, this, std::placeholders::_1));
 }
 
+// todo: is it still necessary
 CostmapSubscriber::CostmapSubscriber(
   const rclcpp::Node::WeakPtr & parent,
   const std::string & topic_name)
 : topic_name_(topic_name)
 {
   auto node = parent.lock();
+  logger_ = node->get_logger();
   costmap_sub_ = node->create_subscription<nav2_msgs::msg::Costmap>(
     topic_name_,
     rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
     std::bind(&CostmapSubscriber::costmapCallback, this, std::placeholders::_1));
+  costmap_update_sub_ = node->create_subscription<nav2_msgs::msg::CostmapUpdate>(
+    topic_name_ + "_updates",
+    rclcpp::QoS(rclcpp::KeepLast(costmapUpdateQueueDepth)).transient_local().reliable(),
+    std::bind(&CostmapSubscriber::costmapUpdateCallback, this, std::placeholders::_1));
 }
 
+// todo: change to const?
 std::shared_ptr<Costmap2D> CostmapSubscriber::getCostmap()
 {
   if (!costmap_received_) {
     throw std::runtime_error("Costmap is not available");
   }
-  toCostmap2D();
+  if (costmap_msg_ != nullptr) {
+    processCurrentCostmapMsg();
+  }
   return costmap_;
 }
 
-void CostmapSubscriber::toCostmap2D()
+void CostmapSubscriber::costmapCallback(const nav2_msgs::msg::Costmap::SharedPtr msg)
 {
-  auto current_costmap_msg = std::atomic_load(&costmap_msg_);
-
-  if (costmap_ == nullptr) {
-    costmap_ = std::make_shared<Costmap2D>(
-      current_costmap_msg->metadata.size_x, current_costmap_msg->metadata.size_y,
-      current_costmap_msg->metadata.resolution, current_costmap_msg->metadata.origin.position.x,
-      current_costmap_msg->metadata.origin.position.y);
-  } else if (costmap_->getSizeInCellsX() != current_costmap_msg->metadata.size_x ||  // NOLINT
-    costmap_->getSizeInCellsY() != current_costmap_msg->metadata.size_y ||
-    costmap_->getResolution() != current_costmap_msg->metadata.resolution ||
-    costmap_->getOriginX() != current_costmap_msg->metadata.origin.position.x ||
-    costmap_->getOriginY() != current_costmap_msg->metadata.origin.position.y)
   {
-    // Update the size of the costmap
-    costmap_->resizeMap(
-      current_costmap_msg->metadata.size_x, current_costmap_msg->metadata.size_y,
-      current_costmap_msg->metadata.resolution,
-      current_costmap_msg->metadata.origin.position.x,
-      current_costmap_msg->metadata.origin.position.y);
+    std::lock_guard<std::mutex> lock(costmap_msg_mutex_);
+    costmap_msg_ = msg;
+  }
+  if (!costmap_received_) {
+    costmap_ = std::make_shared<Costmap2D>(
+      msg->metadata.size_x, msg->metadata.size_y,
+      msg->metadata.resolution, msg->metadata.origin.position.x,
+      msg->metadata.origin.position.y);
+
+    processCurrentCostmapMsg();
+
+    costmap_received_ = true;
   }
+}
+
+void CostmapSubscriber::costmapUpdateCallback(
+  const nav2_msgs::msg::CostmapUpdate::SharedPtr update_msg)
+{
+  if (costmap_received_) {
+    if (costmap_msg_ != nullptr) {
```

Comment:
```suggestion
    if (costmap_msg_) {
```

Is the same :-)

Reply:
resolved with: 33bf135588a90c6db0f4745b1155de7e00941f2c

Please resolve this conversation if you can see these changes now.